### PR TITLE
OTT-401 :: Supported price floor feature 

### PR DESCRIPTION
**Pull request by pm-nikhil-vaidya**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: floors/floors.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+// IsRequestEnabledWithFloor will check if floors is enabled in request
+func IsRequestEnabledWithFloor(Floors *openrtb_ext.PriceFloorRules) bool {
+	return Floors != nil && Floors.Enabled != nil && *Floors.Enabled
+}
+
+// UpdateImpsWithFloors will validate floor rules, based on request and rules prepares various combinations
+// to match with floor rules and selects appripariate floor rule and update imp.bidfloor and imp.bidfloorcur
+func UpdateImpsWithFloors(floorExt *openrtb_ext.PriceFloorRules, request *openrtb2.BidRequest, conversions currency.Conversions) []error {
+	var floorErrList []error
```
Comment (Pubmatic-Dhruv-Sonone) [inline floors/floors.go:41]: Can we combine this in single var block?

File: config/config.go [hunk 2]
```diff
@@ -1201,6 +1209,10 @@ func SetupViper(v *viper.Viper, filename string) {
 	v.SetDefault("gdpr.tcf2.purpose_one_treatment.access_allowed", true)
 	v.SetDefault("gdpr.tcf2.special_feature1.enforce", true)
 	v.SetDefault("gdpr.tcf2.special_feature1.vendor_exceptions", []openrtb_ext.BidderName{})
+	v.SetDefault("price_floors.enabled", false)
```
Comment (pm-nikhil-vaidya) [inline config/config.go:1212]: Add warning in validate function if floors is enabled

**Review by pm-viral-vala — CHANGES_REQUESTED** [review]

File: exchange/enforce_floors.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package exchange
```
Comment (pm-nikhil-vaidya) [inline exchange/enforce_floors.go:1]: rename file to floors.go

File: floors/rule.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+			value = getpbadslot(imp)
+		}
+		ruleKeys = append(ruleKeys, value)
+	}
+	return ruleKeys
+}
+
+func getDeviceType(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if isMobileDevice(request.Device.UA) == true {
```
Comment (pm-nikhil-vaidya) [inline floors/rule.go:135]: remove comparison with true

File: exchange/enforce_floors.go (hunk 3 above)
Comment (pm-nikhil-vaidya) [reply exchange/enforce_floors.go:1]: done

File: config/config.go (hunk 2 above)
Comment (pm-nikhil-vaidya) [reply config/config.go:1212]: PriceFloors.Enabled will enforce floor feature which is still under development.

Comment (pm-nikhil-vaidya) [reply config/config.go:1212]: Added the message

File: exchange/exchange.go [hunk 5]
```diff
@@ -89,14 +91,14 @@ type BidIDGenerator interface {
...
-}
-
 func (big *bidIDGenerator) Enabled() bool {
 	return big.enabled
 }
 
+type bidIDGenerator struct {
+	enabled bool
+}
+
```
[the remark this answers is not in the document]
Comment (pm-nikhil-vaidya) [reply exchange/exchange.go:101]: Done

File: exchange/exchange.go [hunk 6]
```diff
@@ -142,7 +144,12 @@ func NewExchange(adapters map[openrtb_ext.BidderName]adaptedBidder, cache prebid
 		},
 		bidIDGenerator: &bidIDGenerator{cfg.GenerateBidID},
 		gvlVendorIDs:   infos.ToGVLVendorIDMap(),
-		trakerURL:      cfg.TrackerURL,
+		floor: &floors.FloorConfig{
```
[the remark this answers is not in the document]
Comment (pm-nikhil-vaidya) [reply exchange/exchange.go:147]: Done

File: exchange/exchange.go [hunk 7]
```diff
@@ -187,6 +194,7 @@ type BidderRequest struct {
 }
 
 func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *DebugLog) (*openrtb2.BidResponse, error) {
+
```
[the remark this answers is not in the document]
Comment (pm-nikhil-vaidya) [reply exchange/exchange.go:197]: Done

File: floors/rule.go (hunk 4 above)
Comment (pm-nikhil-vaidya) [reply floors/rule.go:135]: done

File: exchange/exchange.go [hunk 8]
```diff
@@ -285,7 +304,21 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
...
+				enforceDealFloors = requestExt.Prebid.Floors.Enforcement.FloorDeals && e.floor.EnforceDealFloor()
+			}
+			adapterBids, rejections = EnforceFloorToBids(r.BidRequestWrapper.BidRequest, adapterBids, conversions, enforceDealFloors)
+			for _, message := range rejections {
+				errs = append(errs, errors.New(message))
+			}
+		}
+
 		adapterBids, rejections := applyAdvertiserBlocking(r.BidRequestWrapper.BidRequest, adapterBids)
+
```
[the remark this answers is not in the document]
Comment (pm-nikhil-vaidya) [reply exchange/exchange.go:321]: Done

File: floors/floors.go (hunk 1 above)
Comment (pm-nikhil-vaidya) [reply floors/floors.go:55]: done

File: exchange/exchange.go [hunk 9]
```diff
@@ -240,6 +248,20 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
...
+	if e.floor != nil && e.floor.Enabled() && floors.IsRequestEnabledWithFloor(requestExt.Prebid.Floors) {
+		errs = floors.UpdateImpsWithFloors(requestExt.Prebid.Floors, r.BidRequestWrapper.BidRequest, conversions)
+		r.BidRequestWrapper.BidRequest.Ext, err = json.Marshal(requestExt)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		updatedBidReq, _ := json.Marshal(r.BidRequestWrapper.BidRequest)
+		glog.Infof("\n Updated Floor Request after parsing floors = %v", string(updatedBidReq))
+	}
+
```
[the remark this answers is not in the document]
Comment (pm-nikhil-vaidya) [reply exchange/exchange.go]: Will fix this with another refactoring ticket

File: exchange/exchange.go [hunk 10]
```diff
@@ -285,7 +304,21 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
...
+			var enforceDealFloors bool
+			if requestExt.Prebid.Floors.Enforcement != nil {
+				enforceDealFloors = requestExt.Prebid.Floors.Enforcement.FloorDeals && e.floor.EnforceDealFloor()
+			}
+			adapterBids, rejections = EnforceFloorToBids(r.BidRequestWrapper.BidRequest, adapterBids, conversions, enforceDealFloors)
+			for _, message := range rejections {
+				errs = append(errs, errors.New(message))
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (pm-nikhil-vaidya) [reply exchange/exchange.go:313]: We will fix this with another refactoring ticket

**Pull request merged**